src/interface/blas2_interface.hpp

```diff
@@ -55,67 +55,85 @@ typename Executor::policy_t::event_t _gemv_impl(
-  const index_t nRowsWG =
-      (_nRowsWG == 0) ? ((data_layout_t::is_col_major()) ? localSize : 1)
-                      : std::min(M, _nRowsWG);
-  const index_t nColsWG =
-      (_nColsWG == 0) ? ((data_layout_t::is_col_major()) ? localSize : N)
-                      : std::min(N, _nColsWG);
+  auto mA = make_matrix_view<col_major>(ex, _mA, _M, _N, _lda);
+  auto vx = make_vector_view(ex, _vx, _incx, x_vector_size);
+  auto vy = make_vector_view(ex, _vy, _incy, _M);
+
+  const index_t Ns_in_WG = local_size;
+  const index_t Ms_in_WG = local_size;
+  
   const index_t scratchPadSize =
-      (_localSize == 0) ? localSize : _scratchPadSize;
+      (_localSize == 0) ? 0 : _scratchPadSize;
 
-  const index_t nWGPerCol = (N - 1) / nColsWG + 1;
-  const index_t nWGPerRow = (M - 1) / nRowsWG + 1;
-  const index_t globalSize = localSize * nWGPerRow * nWGPerCol;
+  const index_t WGs_per_NC = is_transposed
+                                    ? (_N - 1) / Ns_in_WG + 1
+                                    : (_M - 1) / Ms_in_WG + 1;
+  const index_t WGs_per_C = is_transposed
+                                    ? (_M - 1) / Ms_in_WG + 1
+                                    : (_N - 1) / Ns_in_WG + 1;
 
-  const index_t scratchSize =
-      (data_layout_t::is_col_major())
-          ? nWGPerCol
-          : (((scratchPadSize == 0) ? std::min(N, localSize) : 1) * nWGPerCol);
+  index_t kernel_scratch_size = 
+                    (scratchPadSize == 0)
+                      ? local_size + (is_transposed ? (cache_capacity + 1) * local_size : 0)
+                      : scratchPadSize;
 
-  auto valT1 = blas::make_sycl_iterator_buffer<element_t>(M * scratchSize);
-  // this is column major
-  auto mat1 =
-      make_matrix_view<row_major>(ex, valT1, M, scratchSize, scratchSize);
+  const index_t output_scratch_size = WGs_per_C;
 
-  if (data_layout_t::is_col_major()) {
-    auto gemvC =
-        make_Gemv_Col(mat1, mA, vx, nWGPerRow, nWGPerCol, scratchPadSize);
-    ret = concatenate_vectors(
-        ret, ex.execute(gemvC, localSize, globalSize, scratchPadSize));
-  } else {
-    auto gemvR = make_Gemv_Row<interLoop>(mat1, mA, vx, nWGPerRow, nWGPerCol,
-                                          scratchPadSize);
-    ret = concatenate_vectors(
-        ret, ex.execute(gemvR, localSize, globalSize, scratchPadSize));
-  }
+  // Leading dimension for partial dot products matrix
+  auto ld = !is_transposed ? _M : _N;
+  auto dot_products_size = ld * output_scratch_size;
+
+  auto dot_products_buffer = 
+        blas::make_sycl_iterator_buffer<element_t>(dot_products_size);
+  auto dot_products_matrix = 
+        make_matrix_view<col_major>(ex, dot_products_buffer, ld, output_scratch_size, ld);
+
+  // Global size for shared-mem-kernel
+  const index_t global_size = local_size * WGs_per_C * WGs_per_NC;
+
+  // Global size for no-shared-mem kernel
+  //const index_t global_size = roundUp(dot_products_size, local_size);
```

Comment:
Yeah pretty much. Basically right now the user cannot choose whether the shared-memory or non-shared-memory kernel runs. I made shared-memory the default, but the calculation for `global_size` is different for the non-shared memory kernel, so I left this in here in case someone was adding this functionality in the future. It would require quite a lot of changes to the interface though I think.

Reply:
I've deleted it. The non-shared-mem kernel will remain, but there is no way to call it currently without bigger changes to the interface.